### ISSUE TITLE
Garbage collector should clean up the partition catalog records too

### ISSUE THREAD

**Issue by carols10cents** [influxdata/influxdb_iox#8417]
Related to: 

- https://github.com/influxdata/influxdb_iox/issues/8066
- https://github.com/influxdata/idpe/issues/17476

We've been operating under the assumption that old partition records that don't have a partition hash ID will "age out" of the catalog as the data gets to the retention period and gets cleaned up by the garbage collector process... but [the garbage collector process only cleans up Parquet file records](https://github.com/influxdata/influxdb_iox/blob/81b5d80a91423f54f625a36f29b97b5f89d504aa/garbage_collector/src/parquetfile/deleter.rs#L18-L24), not the parent partition records.

To prevent the partition table from growing infinitely and to enable transition to a world in which all partition records have a hash ID, the garbage collector should clean up partition records that no longer have Parquet file records after those are deleted. I see a few potential ways of doing this:

1. Use the deleted Parquet files to find partitions to delete:
  - Change [`ParquetFileRepo::delete_old_ids_only`](https://github.com/influxdata/influxdb_iox/blob/81b5d80a91423f54f625a36f29b97b5f89d504aa/iox_catalog/src/postgres.rs#L1597-L1620) to return the `parquet_file` `partition_id` and `partition_hash_id` columns in addition to the `id` column (and change the method name)
  - Query for any remaining Parquet file records with those `partition_id` or `partition_hash_id` values. If none exist, delete the partition record with those `partition_id` or `partition_hash_id` values.
2. Query separately for old partitions that have no Parquet files
  - Partition tables have [a `new_file_at` column](https://github.com/influxdata/influxdb_iox/blob/81b5d80a91423f54f625a36f29b97b5f89d504aa/data_types/src/partition.rs#L388-L389). We could query for partitions where `new_file_at` is less than a cutoff time ago ([much like the Parquet file query](https://github.com/influxdata/influxdb_iox/blob/81b5d80a91423f54f625a36f29b97b5f89d504aa/garbage_collector/src/parquetfile/deleter.rs#L16)) AND there are no Parquet file records associated with that partition

I think we have to at least do number 2 to clean up the partition records whose Parquet file records have already been deleted by the garbage collector.

Is my logic correct here? Is there anything I'm missing?

**Comment by domodwyer** [comment]
Yeah I think deleting the partitions after a set amount of time is the way to go (same strategy as the parquet files).

I think the challenge to this work will be in identifying bits of code where the assumption is that partitions live forever (I know the ingester definitely does, and probably the querier/compactor too). It might be that a delay helps "hide" the problem, but for the ingester's thats probably not good enough (smaller number of them, easy to take down all ingest with a couple of well timed crashes)

**Comment by philjb** [comment]
see also https://github.com/influxdata/idpe/issues/17994

**Comment by philjb** [comment]
> to enable transition to a world in which all partition records have a hash ID

you may never get to this world. A database/partition could have infinite retention for the data.

**Comment by carols10cents** [comment]
i need 😬 in the github reactions
